backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/CodeGenerator.kt

```diff
@@ -935,12 +891,33 @@ internal class FunctionGenerationContext(val function: LLVMValueRef,
                 )
             }
         }
+
+        // If object is imported - access it via getter function.
+        if (isExternal(irClass)) {
+            val valueGetterName = irClass.objectInstanceGetterSymbolName
+            val valueGetterFunction = LLVMGetNamedFunction(context.llvmModule, valueGetterName) ?:
+                LLVMAddFunction(context.llvmModule, valueGetterName,
+                        functionType(kObjHeaderPtr, false, kObjHeaderPtrPtr))
+            return call(valueGetterFunction!!,
+                        listOf(),
+                        resultLifetime = Lifetime.GLOBAL,
+                        exceptionHandler = exceptionHandler)
+        }
+
         val storageKind = irClass.storageKind(context)
-        if (storageKind == ObjectStorageKind.PERMANENT) {
-            return loadSlot(codegen.getObjectInstanceStorage(irClass, storageKind), false)
+        when (storageKind) {
+            ObjectStorageKind.SHARED -> context.llvm.sharedObjects += irClass
```

Comment:
Somewhat harmful for DCE.
Now global deinitializer zeroes storage for all defined `public` singletons, even for unused ones.

Reply:
Doesn't look like big deal.